src/Raven.Server/Rachis/TimeoutEvent.cs

```diff
@@ -117,27 +117,41 @@ public void DisableTimeout()
             }
         }
 
+        private int _inProgress;
+
         public void ExecuteTimeoutBehavior()
         {
+            var action = _timeoutHappened;
+
+            if (Interlocked.CompareExchange(ref _inProgress, 1, 0) == 1)
```

Comment:
What is the dead lock that you are running into here?
And why not handle that using the already existing `lock (this)` by changing this to be `Monitor.TryEnter` ?

Reply:
p1 is the timer callback (`SwitchToCandidateStateOnTimeout`)
p2 switching the rachis state (`SetNewState`)

p1 enters the lock -> p2 opens write tx -> p2 try to enter the lock but blocked -> p1 tries to open write tx.

Reply:
Not sure how `Monitor.TryEnter` would help.
I need that the execution of the callback to be outside of the lock.